### ISSUE TITLE
parse alerting before loading WAL

### ISSUE THREAD

**Issue by anarcat** [prometheus/prometheus#11486]
### Proposal

Just like we parse the configuration file before the WAL (#7399, #7322), we should parse rules files before loading the WAL.

I have just had a pretty nasty incident where this problem, combined with a systemd unit that would `Restart=on-failure`, lead to a total disk exhaustion after systemd infinitely retried to reload a config broken before the weekend (well, two days before, but it seems no one noticed before the weekend).

The gory details of that incident are in:

https://gitlab.torproject.org/tpo/tpa/team/-/issues/40939

I have also filed this bug against the Debian package to fix the systemd unit:

https://bugs.debian.org/1022724

**Comment by SuperQ** [comment]
I think the systemd unit starts on failure because it always exits non-zero.

**Comment by dswarbrick** [comment]
One of the cornerstones of Prometheus is that you should cross-monitor if you have multiple instances, i.e. configure each instance to monitor every other instance (and alert on failure).

I highly recommend using a CI method for managing / deploying configs, which should include the following steps before deploying:
* promtool check config /path/to/prometheus.yml
* promtool check rules  /path/to/rules/*.yml

**Comment by anarcat** [comment]
yes, you can see in the aforementioned bug report that i took multiple measures to keep this catastrophe from happening again. thankfully we don't (yet) use prometheus as our core alerting software (stuck in icinga land rn), so this wasn't *that* bad, but still...

we now have CI checking those things *and* i made a fix to the debian package to keep it away from the restart death spiral as well.

i just figured it might be good to fix this on prometheus' side as well, and @SuperQ actually asked for a bug report here as well.

**Comment by dswarbrick** [comment]
@anarcat Maybe... I wasn't aware that `promtool check config` also checks rules. I don't recall it always doing that. The examples I provided are from a CI config which I implemented about 3-4 years ago as part of a GitLab CI config. Having the `promtool check rules` as an explicit, separate step made it clearer to the users who were submitting merge requests why the pipeline was failing.

**Comment by roidelapluie** [comment]
This sounds like a good idea

**Comment by beorn7** [comment]
Hello from the bug scrub!

Essentially, we want to do the config parsing before doing any expensive steps during startup. Is that the right understanding?

Sadly, nobody tackled this in the last two years, but we would welcome PRs.

**Comment by dennisme** [comment]
Hey this sounds like it could save someone time in the future waiting for wal replay if there is a config issue deployed prior to the restart. Im game to take this one @beorn7.

**Comment by dennisme** [comment]
Thanks @beorn7

**Comment by suryaaprakassh** [comment]
Hi im new to this , i would like to work on this issue, looking for some guidance

**Comment by beorn7** [comment]
@suryaaprakassh this is already assigned to @dennisme .
@dennisme , do you plan to work on this any time soon?

**Comment by beorn7** [comment]
With no response from @dennisme , I'll unassign this.

**Comment by marcodebba** [comment]
Hi! I'm currently looking for an open source project to start contributing to, and this issue on Prometheus seems like a great fit.
Would it be okay if I take on this ? Let me know if there are any guidelines I should follow

**Comment by beorn7** [comment]
We might deal with this issue during the ContribFest Wednesday next week (2025-04-02).

@marcodebba if this is still untouched after Wednesday, you are welcome to work on this.

**Comment by beorn7** [comment]
We have two folks from the contrib fest working on this.

**Comment by marcodebba** [comment]
I Was at contribfest thos morning and i work with the person who got it. No problem ;-)

**Comment by marcodebba** [comment]
Working on this

**Comment by marcodebba** [comment]
Hi @beorn7 ,
If i've understood correctly the problem, the application should stop if the config file prometheus.yml doesn't exist, before "the other expensive steps during startup". If yes, the check should  be extended to the file detailed into the configfile, when it exists. Right ?

**Comment by beorn7** [comment]
At the contribfest, we had gone pretty far nailing down what has to be done here. But I don't remember the detail. @marcodebba were you at the same table when we discussed it? Who (else) was there?

**Comment by marcodebba** [comment]
Grazie per gli importanti dettagli. Ero proprio accanto al ragazzo che aveva iniziato a vedere il codice insieme a te per capire meglio l'issue.

**Comment by marcodebba** [comment]
added a func parseRule called in a new section which loads and check the rules files. Hope to send something next week

**Comment by marcodebba** [comment]
I'm sending a PR with the new check. Fixing all the test failed

**Comment by marcodebba** [comment]
Sent a PR, waiting for review

**Comment by subhramit** [comment]
@marcodebba hey, are you still working on it? If not, I was thinking of taking over and finishing it.

**Comment by marcodebba** [comment]
Hi @subhramit , yes, no problem for me. Is a very simple issue, and i've finished it, but for my job now is very difficult a simple PR . So , it's yours.

**Comment by subhramit** [comment]
> Hi [@subhramit](https://github.com/subhramit) , yes, no problem for me. Is a very simple issue, and i've finished it, but for my job now is very difficult a simple PR . So , it's yours.

Thanks. I have added you as an author in commit https://github.com/prometheus/prometheus/pull/16601/commits/304d1c904849888f9a742153f962f0e7870a82c2 and filed a [fresh PR](https://github.com/prometheus/prometheus/pull/16462) as I did not want to work on a main branch (and perhaps would not have permission to work on your fork).
